// marko/dist/runtime/RenderResult.js
"use strict";var domInsert = require("./dom-insert");


function getRootNode(el) {
  var cur = el;
  while (cur.parentNode) cur = cur.parentNode;
  return cur;
}

function getComponentDefs(result) {
  var componentDefs = result.b_;

  if (!componentDefs) {
    throw Error("No component");
  }
  return componentDefs;
}

function RenderResult(out) {
  this.out = this.z_ = out;
  this.b_ = undefined;
}

module.exports = RenderResult;

var proto = RenderResult.prototype = {
  getComponent: function () {
    return this.getComponents()[0];
  },
  getComponents: function (selector) {
    if (this.b_ === undefined) {
      throw Error("Not added to DOM");
    }

    var componentDefs = getComponentDefs(this);

    var components = [];

    componentDefs.forEach(function (componentDef) {
      var component = componentDef.h_;
      if (!selector || selector(component)) {
        components.push(component);
      }
    });

    return components;
  },

  afterInsert: function (host) {
    var out = this.z_;
    var componentsContext = out.b_;
    if (componentsContext) {
      this.b_ = componentsContext.A_(host);
    } else {
      this.b_ = null;
    }

    return this;
  },
  getNode: function (host) {
    return this.z_.B_(host);
  },
  getOutput: function () {
    return this.z_.C_();
  },
  toString: function () {
    return this.z_.toString();
  },
  document: typeof document === "object" && document };


Object.defineProperty(proto, "html", {
  get: function () {
    // eslint-disable-next-line no-constant-condition





    return this.toString();
  } });


Object.defineProperty(proto, "context", {
  get: function () {
    // eslint-disable-next-line no-constant-condition





    return this.z_;
  } });


// Add all of the following DOM methods to Component.prototype:
// - appendTo(referenceEl)
// - replace(referenceEl)
// - replaceChildrenOf(referenceEl)
// - insertBefore(referenceEl)
// - insertAfter(referenceEl)
// - prependTo(referenceEl)
domInsert(
proto,
function getEl(renderResult, referenceEl) {
  return renderResult.getNode(getRootNode(referenceEl));
},
function afterInsert(renderResult, referenceEl) {
  return renderResult.afterInsert(getRootNode(referenceEl));
});
